### PR TITLE
[WIP] Parallel timeout for fetching PPD files.

### PR DESCRIPTION
**Pull request by pranjanpr**
Issue https://github.com/OpenPrinting/cups/issues/65
When some of the PPD generators took more then 60 seconds to get PPD files then `cups` expires the scheduler. 

To solve the issue, created a separate thread along with the main thread. Main thread waits for the timeout and the separate thread calls `cupsFileGets()` to get the PPD files. If it is not able to fetch it within timeout, the main thread terminates the separate thread.

@zdohnal @michaelrsweet Please review it. Thanks !

### CODE REVIEW

File: scheduler/cups-driverd.cxx [hunk 1]
```diff
@@ -21,7 +21,10 @@
 #include <cups/transcode.h>
 #include <cups/ppd-private.h>
 #include <ppdc/ppdc.h>
+#include <pthread.h>
```
Comment (michaelrsweet) [inline scheduler/cups-driverd.cxx:24]: As the pthread code in CUPS is currently *optional*, you need to make running the driver interfaces in parallel optional (and the pthread include optional) based on the HAVE_PTHREAD preprocessor define.

File: scheduler/cups-driverd.cxx [hunk 2]
```diff
@@ -137,6 +146,7 @@ static const char * const PPDTypes[] =	/* ppd-type values */
 			  "drv",
 			  "archive"
 			};
+static char             *PPDVal = NULL; /* Return value of cupsFileGets() */
```
Comment (michaelrsweet) [inline scheduler/cups-driverd.cxx:149]: Globals and threading do not go together well...

File: scheduler/cups-driverd.cxx [hunk 3]
```diff
@@ -2914,3 +2926,81 @@ regex_string(const char *s)		/* I - String to compare */
...
+/*
+ * 'fetch_ppd()' - Thread function to fetch PPD files.
+ */
+
+void *
+fetch_ppd(void *input)
+{
+
+  ppd_gen_t *alias = (ppd_gen_t *)input;
+  PPDVal = cupsFileGets(alias -> fp, alias -> buf, alias -> buflen);
```
Comment (michaelrsweet) [inline scheduler/cups-driverd.cxx:2940]: This does not look safe or at all what I'd expect for a threaded reader...

**Review by michaelrsweet — CHANGES_REQUESTED** [review]
This doesn't appear to do what we need it to do.

In order for this change to work, we need to run the driver interface programs in the background (in separate threads, or just poll'ing with the cupsFileNumber's which might be simpler and it will be easier to time them out together) and then limit the total amount of time they can run...

**Comment by pranjanpr** [comment]
> This doesn't appear to do what we need it to do.
> 
> In order for this change to work, we need to run the driver interface programs in the background (in separate threads, or just poll'ing with the cupsFileNumber's which might be simpler and it will be easier to time them out together) and then limit the total amount of time they can run...

Okay so the idea is to divide the ppd generators into different available threads and use `poll.h` (instead of pthread which I have done currently) in every thread such that it does not exceed timeout. I will first try to implement `poll.h` as it would be universal in every scenario such as when the CPU won't be having multiple threads.

**Comment by pranjanpr** [comment]
@michaelrsweet 
SIr, I have added `poll.h` and removed the former `pthread.h` for creating timeout for PPD descriptors. Please review it. 
Thanks !

**Review by michaelrsweet — CHANGES_REQUESTED** [review]
No, that's not running the driver interfaces in parallel, that's just timing out a single driver interface. The goal is to run all of the driver interfaces, poll multiple file descriptors, and limit the total run time.

**Comment by zdohnal** [comment]
The original issue is closed, so the enhancement is no longer required.

Thank you for the PR nevertheless!

**Pull request closed without merging**